File: src/main/java/cn/enaium/joe/stage/ConfigDialog.java
/*
 * Copyright 2022 Enaium
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.enaium.joe.stage;

import cn.enaium.joe.MainFX;
import cn.enaium.joe.annotation.NoUI;
import cn.enaium.joe.config.Config;
import cn.enaium.joe.config.value.*;
import cn.enaium.joe.gui.panel.ColumnPane;
import cn.enaium.joe.util.LangUtil;
import cn.enaium.joe.util.MessageUtil;
import javafx.geometry.Pos;
import javafx.scene.control.*;

import java.lang.reflect.Field;

/**
 * @author Enaium
 * @since 1.3.0
 */
public class ConfigDialog extends Dialog {
    public ConfigDialog(Config config) {
        super(LangUtil.i18n("menu.config") + ":" + config.getName(), new ColumnPane() {{
            try {
                int row = 0;
                for (Field declaredField : config.getClass().getDeclaredFields()) {
                    declaredField.setAccessible(true);

                    if (declaredField.isAnnotationPresent(NoUI.class)) {
                        continue;
                    }

                    Object o = declaredField.get(config);

                    if (o instanceof Value) {
                        Value<?> value = (Value<?>) o;
                        Label label = new Label(value.getName());
                        Tooltip.install(label, new Tooltip(value.getDescription()));

                        if (o instanceof StringValue) {
                            StringValue stringValue = (StringValue) o;
                            add(label, new TextField(stringValue.getValue()) {{
                                textProperty().addListener((observableValue, s, t1) -> stringValue.setValue(getText()));
                            }});
                        } else if (o instanceof IntegerValue) {
                            IntegerValue integerValue = (IntegerValue) o;
                            add(label, new Spinner<Integer>(0, 256, integerValue.getValue()) {{
                                valueProperty().addListener((observableValue, integer, t1) -> integerValue.setValue(getValue()));
                            }});
                        } else if (o instanceof EnableValue) {
                            EnableValue enableValue = (EnableValue) o;
                            add(label, new CheckBox() {{
                                setAlignment(Pos.CENTER_RIGHT);
                                setSelected(enableValue.getValue());
                                setOnAction(actionEvent -> enableValue.setValue(isSelected()));
                            }});
                        } else if (o instanceof ModeValue) {
                            add(label, new ComboBox<String>() {{
                                ModeValue modeValue = (ModeValue) o;
                                for (String s : modeValue.getMode()) {
                                    getItems().add(s);
                                    getSelectionModel().select(modeValue.getValue());
                                }
                                setOnAction(actionEvent -> {
                                    modeValue.setValue(getSelectionModel().getSelectedItem());
                                });
                            }});
                        }
                    }
                }
            } catch (IllegalAccessException e) {
                MessageUtil.error(e);
            }
        }});
        setOnCloseRequest(windowEvent -> {
            MainFX.getInstance().config.save();
        });
    }
}

File: ui/src/main/kotlin/cn/enaium/joe/ui/dialog/ConfigDialog.kt
/*
 * Copyright 2022 Enaium
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.enaium.joe.ui.dialog

import cn.enaium.joe.core.config.Config
import cn.enaium.joe.core.config.value.*
import cn.enaium.joe.ui.config.value.KeyCodeValue
import cn.enaium.joe.ui.util.i18n
import javafx.geometry.Pos
import javafx.scene.control.*
import javafx.scene.input.KeyCodeCombination
import javafx.scene.input.KeyCombination
import org.tbee.javafx.scene.layout.MigPane

/**
 * @author Enaium
 * @since 2.0.0
 */
class ConfigDialog(config: Config) : Dialog(i18n("menu.config")) {
    init {
        content = ScrollPane(MigPane("fillx", "[fill][fill]").apply {
            config.javaClass.declaredFields.forEach { field ->
                field.isAccessible = true
                val o = field.get(config)
                if (o is Value<*>) {
                    add(Label(o.name))
                }

                when (o) {
                    is StringValue -> {
                        add(TextField(o.value).apply {
                            setOnInputMethodTextChanged {
                                o.value = text
                            }
                        }, "wrap")
                    }

                    is IntegerValue -> {
                        add(Spinner<Int>(0, Int.MAX_VALUE, o.value).apply {
                            setOnInputMethodTextChanged {
                                o.value = value
                            }
                        }, "wrap")
                    }

                    is EnableValue -> {
                        add(CheckBox().apply {
                            alignment = Pos.CENTER_RIGHT
                            isSelected = o.value
                            setOnAction {
                                o.value = isSelected
                            }
                        }, "wrap")
                    }

                    is ModeValue -> {
                        add(ComboBox<String>().apply {
                            value = o.value
                            o.mode.forEach {
                                items.add(it)
                            }
                            selectionModel.selectedItemProperty()
                                .addListener { _, _, newValue ->
                                    o.value = newValue
                                }
                        }, "wrap")
                    }

                    is KeyCodeValue -> {
                        add(Button(o.value.name).apply {
                            setOnKeyPressed {
                                val modifiers = mutableListOf<KeyCombination.Modifier>()

                                if (it.isAltDown) {
                                    modifiers.add(KeyCombination.ALT_DOWN)
                                }

                                if (it.isControlDown) {
                                    modifiers.add(KeyCombination.CONTROL_DOWN)
                                }

                                if (it.isShiftDown) {
                                    modifiers.add(KeyCombination.SHIFT_DOWN)
                                }

                                if (!it.code.isModifierKey) {
                                    o.value = KeyCodeCombination(it.code, *modifiers.toTypedArray())
                                    text = o.value.name
                                }
                            }
                        }, "wrap")
                    }
                }
            }
        }).apply {
            isFitToWidth = true
            isFitToHeight = true
        }
    }
}